### ISSUE TITLE
Your Apple ID or password is incorrect

### ISSUE THREAD

**Issue by Tabstone** [Dadoum/SideloadIPA#2]
send M1 but "Your Apple ID or password is incorrect"

**Comment by Dadoum** [comment]
I am aware of this issue actually. 
Will explain here first why Cydia Impactor stopped working and then why it is so hard to made a good replacement:
To install an IPA, you must sign it, and to sign it from the ground up, there are 3 steps:
 - Login
 - Submit a certificate 
 - Sign the IPA with it.

The problem is located at the login step as you can see.
Before, (Xcode <7.3) the login server is idmsa.apple.com. Get more details [here](https://raw.githubusercontent.com/LeanVel/iInject/master/genProvisioningProfileFree.sh) about IDMSA login. Quickly, this was a short identification process that requires just the email and the password.
But now, for all free accounts, you must use another login server, that is named gsa.apple.com.
It uses an SRP protocol, a well known protocol to exchange safely passwords. Yes, SRP is correctly implemented in SideloadIPA. The problem here is that Apple uses a modified version of the protocol. They **do** use SRP, but you needs to add some informations in your request. You need an email, a password and machine authentications stuff. There are the machine information dictionart which are made of the time, locale... but also of 3 fields [X-Apple-]MD, MD-I, MD-RINFO, and a machine headers that are made of a GUID, MD-LU, and machine details. 
These fields are not made locally. They also requires another step before that is named **machine-provisioning**. This is the most difficult step because there is almost no information about its architecture. This step is itself made of 2 steps:
 - A first empty request with some http headers (such as the GUID and the machine information)
 - From the server answer, that is only a pack of bytes, generate CPD, that are another meaningless bytes.
 - From the last server answer, parse MD-LU, MD, MD-I (MD-RINFO is explicitly written).

I am actively trying to reverse the provisioning step. These fields cannot be incorrect or the request will never succeed. If you fail to create the cpd, the request will not succeed. And if you put real fields, they expires in 90 seconds, and after, these fields will return *Your Apple ID or password is incorrect*.
AltStore and ReProvision does not request provision from the ground up, they use Apple libraries that are available for:
 - macOS & iOS (AuthKit) (x86, ARMv7 and ARMv8)
 - Windows (AuthKitWin): this is the one that I am trying to reproduce (x86 only)
 - Linux (libStoreServicesCore.so for Apple Music for Android) (armv7 and armv8 only)

**TLDR:** Password is not incorrect in fact, this is in development, and Apple error messages are unhelpful.

**Comment by Tabstone** [comment]
I found that in the process of generating M1 by the SRP protocol, the password is encrypted again by apple corecrypto PBKDF2, so the password used in the calculation of M1 is incorrect when you submit it to SRP

**Comment by Dadoum** [comment]
I will modify this behavior in my upcoming (it will surely take time) library.

**Comment by Tabstone** [comment]
In fact, I want to use java to implement the process, but it is not very friendly to the pre-processing of passwords

**Comment by Dadoum** [comment]
I think you could check out the Apple's Java Code for Apple Music app. 
Here is a decompiled version, the native libraries are place in the lib folder in the root of the zip file
[here](https://gofile.io/d/EUQGug)

**Comment by Tabstone** [comment]
This does not solve my problem. The reason why the current password is incorrect should be the processing of the password in corecrypto. What I may need is a corecrypto library

**Comment by Dadoum** [comment]
The thing is that the password, even if it is correct, will not be processed and treated as incorrect if OTP (machine provisioning fields) does not match with GUID and informations, or if OTP has expired. Without touching to these fields, you can't try.

**Comment by Tabstone** [comment]
For the SRP protocol, those device information contents do not participate in the M1 password calculation

**Comment by Dadoum** [comment]
yes they are not in the M1, but even if you return the right M1, if information are incorrect, Apple will not continue SRP.

**Comment by Tabstone** [comment]
In fact, no device information is required for gsa login [login](https://bbs.pediy.com/thread-221012%2520.htm)

**Comment by Tabstone** [comment]
In subsequent sessions, the device information and signature are required for the packet header, but it does not seem to be necessary for gsa login.

**Comment by Dadoum** [comment]
If you find a way to get GSA without it, can you send a code snippet, because with all my tests, Apple will never let you login without the correct headers at least.

**Comment by Dadoum** [comment]
(I has been in contact with someone that succeeded to get M2 from Apple Servers, and he got Incorrect password error too, and he fixed it by poking iTunes dll)

**Comment by Tabstone** [comment]
Do you know how to use corecrypto Library in Linux

**Comment by Dadoum** [comment]
Apple Music android app contains some shared objects libs for ARM64 and ARMv7 but it requires libandroid.
CoreCrypto's SRP follows SRP 6a standards anyway, so if you just need SRP, there are good open source libraries.

**Comment by Dadoum** [comment]
And actually corecrypto is open sourced so if you need its source code it is available anywhere on the web.

**Comment by Tabstone** [comment]
It may be that the ability is not enough, I can't compile a dynamic library that can be used by Linux

**Comment by Tabstone** [comment]
After testing, no device AUKIT does not affect the login behavior of GSA. However, the new API replaces myacinfo with AUKIT, so it is unable to use other APIs after login, and the session will be prompted to expire
